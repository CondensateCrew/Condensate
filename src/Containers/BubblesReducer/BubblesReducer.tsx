import React, {useState, useEffect} from 'react';
import { Redirect } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { addChosenWord, removeAllWords } from 'redux/actions';

interface Props {
  chosenWords: string[]
}

const BubblesReducer: React.FC<Props> = ({ chosenWords }) => {
  const dispatch = useDispatch();
  const [ isRedirected, setIsRedirected ] = useState<boolean>(false);
  const [ isDisabled, setIsDisabled ] = useState<boolean>(false);
  const [ count, setCount ] = useState<number>(0);
  const [ bubbles, setBubbles ] = useState<string[]>([]);
  const [ reducedBubbles, setReducedBubbles ] = useState<string[]>([]);

  useEffect(() => {
    setBubbles(chosenWords.slice(count, count + 3));
    setIsDisabled(false);
  }, [ chosenWords, count ]);

  const goToNextTriple = (event: React.MouseEvent):void => {
    const currentTarget = event.target as HTMLElement;
    updateElements(currentTarget);
    setIsDisabled(true);
    setTimeout(() => {
      setReducedBubbles([...reducedBubbles, currentTarget.id]);
      setCount(count + 3);
    }, 1500);
  };

  useEffect(() => {
    if (reducedBubbles.length === 3) {
      dispatch(removeAllWords());
      reducedBubbles.map((bbl: string) => dispatch(addChosenWord(bbl)));
      setIsRedirected(true);
    }
  }, [reducedBubbles, dispatch]);

  const bubbleEls: JSX.Element[] = bubbles.map((bubble: string, i: number) => (
    <p
      key={`${bubble}-${i + 1}`}
      className="bubble">
      <span></span>
      <button id={bubble.toLowerCase()} onClick={goToNextTriple} disabled={isDisabled}>{bubble}</button>
    </p>
  ));

  return (
    <section className="limited-bbls">
      <h3>Choose one</h3>
      <section>
        {!isRedirected
          ? bubbleEls
          : <Redirect to='/instructions/second-rd' />
        }
      </section>
    </section>
  )
}

export const updateElements = (currentTarget: HTMLElement): void => {
  const parent = currentTarget.parentNode as HTMLElement;
  const sibling = currentTarget.previousSibling as HTMLElement;
  parent.classList.add('disappeared');
  sibling.classList.add('bubble-rounds');
  currentTarget.style.animation = 'disappearing 10s';
}

export default BubblesReducer;
