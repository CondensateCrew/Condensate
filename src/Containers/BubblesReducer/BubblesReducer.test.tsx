import React from 'react';
import BubblesReducer from './BubblesReducer';
import { mount } from 'enzyme';
const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch
}));

describe('BubblesReducer component', () => {
  const mockChosenWords = ['Rest', 'Toy', 'Snow', 'Owl', 'Oil', 'Grid', 'Math', 'Leg', 'Men'];

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const wrapper = mount(
      <BubblesReducer chosenWords={mockChosenWords} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
