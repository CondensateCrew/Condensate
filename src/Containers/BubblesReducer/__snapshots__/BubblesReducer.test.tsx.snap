// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BubblesReducer component should match the snapshot 1`] = `
<BubblesReducer
  chosenWords={
    Array [
      "Rest",
      "Toy",
      "Snow",
      "Owl",
      "Oil",
      "Grid",
      "Math",
      "Leg",
      "Men",
    ]
  }
>
  <section
    className="limited-bbls"
  >
    <h3>
      Choose one
    </h3>
    <section>
      <p
        className="bubble"
        key="Rest-1"
      >
        <span />
        <button
          disabled={false}
          id="rest"
          onClick={[Function]}
        >
          Rest
        </button>
      </p>
      <p
        className="bubble"
        key="Toy-2"
      >
        <span />
        <button
          disabled={false}
          id="toy"
          onClick={[Function]}
        >
          Toy
        </button>
      </p>
      <p
        className="bubble"
        key="Snow-3"
      >
        <span />
        <button
          disabled={false}
          id="snow"
          onClick={[Function]}
        >
          Snow
        </button>
      </p>
    </section>
  </section>
</BubblesReducer>
`;
